src/lib/pubkey/ec_group/ec_inner_data.cpp

```diff
@@ -0,0 +1,227 @@
+*/
+
+#include <botan/internal/ec_inner_data.h>
+
+#include <botan/internal/ec_inner_bn.h>
+#include <botan/internal/ec_inner_pc.h>
+#include <botan/internal/pcurves.h>
+
+namespace Botan {
+
+EC_Group_Data::EC_Group_Data(const BigInt& p,
+                             const BigInt& a,
+                             const BigInt& b,
+                             const BigInt& g_x,
+                             const BigInt& g_y,
+                             const BigInt& order,
+                             const BigInt& cofactor,
+                             const OID& oid,
+                             EC_Group_Source source) :
+      m_curve(p, a, b),
+      m_base_point(m_curve, g_x, g_y),
+      m_g_x(g_x),
+      m_g_y(g_y),
+      m_order(order),
+      m_cofactor(cofactor),
+      m_mod_order(order),
+      m_base_mult(m_base_point, m_mod_order),
+      m_oid(oid),
+      m_p_bits(p.bits()),
+      m_order_bits(order.bits()),
+      m_order_bytes((m_order_bits + 7) / 8),
+      m_a_is_minus_3(a == p - 3),
+      m_a_is_zero(a.is_zero()),
+      m_source(source) {
+   if(m_oid.has_value()) {
+      if(const auto id = PCurve::PrimeOrderCurveId::from_oid(m_oid)) {
+         m_pcurve = PCurve::PrimeOrderCurve::from_id(*id);
+      }
+   }
+}
+
+bool EC_Group_Data::params_match(const BigInt& p,
+                                 const BigInt& a,
+                                 const BigInt& b,
+                                 const BigInt& g_x,
+                                 const BigInt& g_y,
+                                 const BigInt& order,
+                                 const BigInt& cofactor) const {
+   return (this->p() == p && this->a() == a && this->b() == b && this->order() == order &&
+           this->cofactor() == cofactor && this->g_x() == g_x && this->g_y() == g_y);
+}
+
+bool EC_Group_Data::params_match(const EC_Group_Data& other) const {
+   return params_match(other.p(), other.a(), other.b(), other.g_x(), other.g_y(), other.order(), other.cofactor());
+}
+
+std::unique_ptr<EC_Scalar_Data> EC_Group_Data::scalar_from_bytes_with_trunc(std::span<const uint8_t> bytes) const {
+   if(m_pcurve) {
+      return std::make_unique<EC_Scalar_Data_PC>(shared_from_this(), m_pcurve->scalar_from_bits_with_trunc(bytes));
+   } else {
+      auto bn = BigInt::from_bytes_with_max_bits(bytes.data(), bytes.size(), m_order_bits);
+      return std::make_unique<EC_Scalar_Data_BN>(shared_from_this(), mod_order(bn));
+   }
+}
+
+std::unique_ptr<EC_Scalar_Data> EC_Group_Data::scalar_from_bytes_mod_order(std::span<const uint8_t> bytes) const {
+   BOTAN_ARG_CHECK(bytes.size() <= 2 * order_bytes(), "Input too large");
+
+   if(m_pcurve) {
+      if(auto s = m_pcurve->scalar_from_wide_bytes(bytes)) {
+         return std::make_unique<EC_Scalar_Data_PC>(shared_from_this(), std::move(*s));
+      } else {
+         throw Invalid_Argument("Failed to complete reduction of scalar bytes");
```

Comment:
That's somewhat inconsistent. Many other methods that return a unique_ptr simply return `nullptr` if something fails but don't throw. Frankly, I'd prefer things to throw over passing a `nullptr` that I have to remember to check for.

Reply:
The behavior between pcurves and the generic code are somewhat inconsistent on this - generic always succeeds no matter how big the input is. Maybe the answer here is to also introduce a (somewhat artificial) limit for the generic case so they behave the same.

Reply:
I changed this so that all failures return nullptr. I also added doc comments indicating which functions can be expected to fail vs which normally always succeed